src/core/server/saved_objects/saved_objects_service.ts

```diff
@@ -177,124 +196,169 @@ export interface SavedObjectsSetupDeps {
+    });
 
     return {
-      getScopedClient: this.clientProvider.getClient.bind(this.clientProvider),
-      setClientFactory: this.clientProvider.setClientFactory.bind(this.clientProvider),
-      addClientWrapper: this.clientProvider.addClientWrapperFactory.bind(this.clientProvider),
-      createInternalRepository: (extraTypes?: string[]) =>
-        createSORepository(adminClient.callAsInternalUser, extraTypes),
-      createScopedRepository: (req: KibanaRequest, extraTypes?: string[]) =>
-        createSORepository(adminClient.asScoped(req).callAsCurrentUser, extraTypes),
+      migrator,
+      clientProvider,
+      getScopedClient: clientProvider.getClient.bind(clientProvider),
+      createScopedRepository: repositoryFactory.createScopedRepository,
+      createInternalRepository: repositoryFactory.createInternalRepository,
     };
   }
 
-  public async start(core: SavedObjectsStartDeps): Promise<InternalSavedObjectsServiceStart> {
-    if (!this.clientProvider) {
-      throw new Error('#setup() needs to be run first');
-    }
-
-    this.logger.debug('Starting SavedObjects service');
+  public async stop() {}
 
-    /**
-     * Note: We want to ensure that migrations have completed before
-     * continuing with further Core startup steps that might use SavedObjects
-     * such as running the legacy server, legacy plugins and allowing incoming
-     * HTTP requests.
-     *
-     * However, our build system optimize step and some tests depend on the
-     * HTTP server running without an Elasticsearch server being available.
-     * So, when the `migrations.skip` is true, we skip migrations altogether.
-     */
-    const cliArgs = this.coreContext.env.cliArgs;
-    const savedObjectsConfig = await this.coreContext.configService
-      .atPath<SavedObjectsConfigType>('migrations')
-      .pipe(first())
-      .toPromise();
-    const skipMigrations = cliArgs.optimize || savedObjectsConfig.skip;
-    await this.migrator!.runMigrations(skipMigrations);
+  private createMigrator(
+    kibanaConfig: KibanaConfigType,
+    savedObjectsConfig: SavedObjectsConfigType,
+    migrationsRetryDelay?: number
+  ): KibanaMigrator {
+    const savedObjectSchemas = new SavedObjectsSchema(this.schemas);
+    const adminClient = this.setupDeps!.elasticsearch.adminClient;
 
-    return {
-      migrator: this.migrator!,
-      clientProvider: this.clientProvider,
-      getScopedClient: this.clientProvider.getClient.bind(this.clientProvider),
-    };
+    return new KibanaMigrator({
+      savedObjectSchemas,
+      savedObjectMappings: this.mappings,
+      savedObjectMigrations: this.migrations,
+      savedObjectValidations: this.validations,
+      logger: this.coreContext.logger.get('migrations'),
+      kibanaVersion: this.coreContext.env.packageInfo.version,
+      config: this.setupDeps!.legacyPlugins.pluginExtendedConfig,
+      savedObjectsConfig,
+      kibanaConfig,
+      callCluster: migrationsRetryCallCluster(
```

Comment:
optional: could you update `migrationsRetryCallCluster` type ? as you touch the files https://github.com/elastic/kibana/blob/fed6d58fc7235a6d2d98e8c00322bd4b998098d3/src/core/server/elasticsearch/retry_call_cluster.ts#L39